dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,261 @@
+
+  ret = avcodec_open2(av_state_->codec_ctx_, av_state_->codec_, nullptr);
+  DALI_ENFORCE(
+    ret == 0,
+    make_string("Could not initialize codec context: ", detail::av_error_string(ret)));
+
+  av_state_->frame_ = av_frame_alloc();
+  DALI_ENFORCE(av_state_->frame_, "Could not allocate the av frame");
+
+  av_state_->packet_ = av_packet_alloc();
+  DALI_ENFORCE(av_state_->packet_, "Could not allocate av packet");
+}
+
+void VideoFileCPU::FindVideoStream() {
+  for (size_t i = 0; i < av_state_->ctx_->nb_streams; ++i) {
+    auto stream = av_state_->ctx_->streams[i];
+    av_state_->codec_params_ = av_state_->ctx_->streams[i]->codecpar;
+    av_state_->codec_ = avcodec_find_decoder(av_state_->codec_params_->codec_id);
+
+    if (av_state_->codec_ == nullptr) {
+      continue;
+    }
+
+    if (av_state_->codec_->type == AVMEDIA_TYPE_VIDEO) {
+        av_state_->stream_id_ = i;
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in a file ", filename_));
+}
+
+VideoFileCPU::VideoFileCPU(const std::string &filename) : 
+  av_state_(std::make_unique<AvState>()),
+  filename_(filename) {
+  av_state_->ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(av_state_->ctx_, "Could not alloc avformat context");
+
+  int ret = avformat_open_input(&av_state_->ctx_, filename.c_str(), nullptr, nullptr);
+  DALI_ENFORCE(
+    ret == 0,
+    make_string("Failed to open video file at path ", filename, "due to ", detail::av_error_string(ret)));
+
+  FindVideoStream();
+  InitAvState();
+  BuildIndex();
+}
+
+void VideoFileCPU::BuildIndex() {
+  // TODO(awolant): Optimize this function for: 
+  //  - CFR
+  //  - index present in the header
+
+  int last_keyframe = -1;
+  while (ReadRegularFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = av_state_->frame_->key_frame;
+    entry.pts = av_state_->frame_->pts;
+    entry.is_flush_frame = false;
+    
+    if (entry.is_keyframe) {
+      last_keyframe = index_.size();
+    }
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  while(ReadFlushFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = false;
+    entry.pts = av_state_->frame_->pts;
+    entry.is_flush_frame = true;
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  Reset();
+}
+
+void VideoFileCPU::CopyToOutput(uint8_t *data) {
+  dest_[0] = data;
+  dest_linesize_[0] = av_state_->frame_->width * Channels();
+
+  int ret = sws_scale(
+    av_state_->sws_ctx_,
+    av_state_->frame_->data,
+    av_state_->frame_->linesize,
+    0,
+    av_state_->frame_->height,
+    dest_,
+    dest_linesize_);
+
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string("Could not convert frame data to RGB: ", detail::av_error_string(ret)));
+}
+
+bool VideoFileCPU::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  int ret = -1;
+  while(av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
```

Comment:
Nitpick (here in some other places):
```suggestion
  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
```

Reply:
Done